R/class-Collections.R

```diff
@@ -0,0 +1,415 @@
+        msg <- "name cannot be empty"
+        errors <- c(errors, msg)
+    }
+
+    # Ensure id columns are valid
+    msg <- validateIdColumns(df, object@recordIdColumn, object@ancestorIdColumns)
+    errors <- c(errors, msg)
+
+    # check that all columns in data are numeric except id columns
+    dataColNames <- names(df)[!names(df) %in% allIdColumns]
+    if (!all(sapply(object@data[, ..dataColNames], is.numeric))) {
+        msg <- sprintf("all columns in data except '%s' must be numeric", paste(allIdColumns, collapse=", "))
+        errors <- c(errors, msg)
+    }
+
+    # collection data should all come from the same entity
+    # using the presence of the period to indicate eda services formatted data
+    if (all(grepl(".", names(df), fixed = TRUE))) {
+        if (uniqueN(veupathUtils::strSplit(names(df)[!names(df) %in% object@ancestorIdColumns], ".", ncol=2, index=1)) > 1) {
+            msg <- paste("All columns must belong to the same entity.")
+            errors <- c(errors, msg)
+        }
+    }
+
+    if (length(errors) == 0) {
+        return(TRUE)
+    } else {
+        return(errors)
+    }
+}
+
+#' Data Collection
+#' 
+#' This class represents a collection of related data. That could be
+#' all of the abundance data for all samples in a dataset at a particular taxonomic rank,
+#' or all pathway abundance data for all samples in a dataset, or something else. The
+#' primary requirement for a collection is that the values of all variables in the
+#' collection can be expressed on the same theoretical range.
+#' @slot name The name of the collection
+#' @slot data A data.frame of data with samples as rows and variables as columns
+#' @slot recordIdColumn The name of the column in the data.frame that contains the record id
+#' @slot ancestorIdColumns A character vector of column names representing parent entities of the recordIdColumn
+#' @slot imputeZero A logical indicating whether NA/ null values should be replaced with zeros.
+#' @slot removeEmptyRecordss A logical indicating whether empty (all NA/ zero) records should be removed.
+#' @name Collection-class
+#' @rdname Collection-class
+#' @export
+setClass("Collection", 
+    slots = c(
+        name = "character",
+        data = "data.table",
+        recordIdColumn = "character",
+        ancestorIdColumns = "character",
+        imputeZero = 'logical',
+        removeEmptyRecords = 'logical'
+    ), prototype = prototype(
+        imputeZero = TRUE,
+        removeEmptyRecords = TRUE
+    ),
+    validity = check_collection
+)
+
+
+check_collections <- function(object) {
+    errors <- character()
+
+    # check that all names are unique
+    if (length(unique(getCollectionNames(object))) != length(object)) {
+        msg <- "collection names must be unique"
+        errors <- c(errors, msg)
+    }
+
+    if (length(object) == 0) {
+        return(TRUE)
+    }
+
+    # check that at least one ancestorIdColumn is shared between collections
```

Comment:
This tests if each collection shares an ancestorId with the first collection. I assumed we'd want them all to have the same ancestorId in their ancestorIdColumns? Or does that matter? If not, comparing with the first misses the case of the first ancestorIds being [1, 2], and the second and third collection ancestorIds being [2, 3] and [3, 4]

Reply:
im trying not to make too many assumptions about the shape of the entity diagram.. particularly since this class could end up used for arbitrary user data. but it does seem like if the ancestor slot is populated for a list of collections, they should come from the same entity diagram? otherwise, you shouldnt be providing ancestor ids.. though now i say it, i think we need to more explicitly support and test cases where ancestor ids arent provided.

Reply:
ok. i think im going to leave better support for things like null ancestor ids to another pr though.